### ISSUE TITLE
SignalR sometimes calls JQuery ready handlers again

### ISSUE THREAD

**Issue by drauch** [dotnet/aspnetcore#20784]
I'm using SignalR on a ASP.NET Core 3.1 page for a little game.

Sometimes though, SignalR performs some kind of synchronous AJAX request and afterwards my JQuery ready handlers are called again. That has some serious implications, because all the event handlers I register in ready events are now registered twice (and it can even get worse, I already experienced up to 6 re-registrations).

Here is a screenshot of the browser's console output of such an occurrence:
![image](https://user-images.githubusercontent.com/1516470/79126523-6458c500-7da0-11ea-9992-d107b3d61f66.png)

Do you know how to overcome the problem?

PS: I just realized that it maybe does so in order to call the ready handler which should re-establish the hub connection?! On the other hand, the console log looks like there are now two connections open from the same browser window - am I correct?

**Comment by BrennanConroy** [comment]
Yes, you have 2 connections open, you should take a look at your connection startup logic, you might be calling it multiple times accidentally.

**Comment by drauch** [comment]
As you can see from the console log output: I have 2 connections open only AFTER SignalR performs this mysterious synchronous AJAX call. Just like any other of my document ready handlers, the one initializing the SignalR connection is also called twice in that scenario which is why I have two connection after the reload.

But why is SignalR performing this synchronous AJAX request + then re-calling all my document ready event handlers once again? That doesn't make sense to me.

**Comment by BrennanConroy** [comment]
SignalR doesn't do any synchronous AJAX requests, and we definitely don't do anything with calling document ready event handlers.

Do you have an app you can share that repros the issue?

**Comment by davidfowl** [comment]
Old signalr does, but new signalr does not.

**Comment by drauch** [comment]
OK, thanks guys. If there is no hidden page reload on connection losses or something like that, I need to dig deeper into the application. Probably a red herring that it occurs in situations in which the SignalR client loses the connection to the hub.

> Old signalr does, but new signalr does not.

I'm using the new SignalR, so the sync AJAX can't be from you and it must come from something else. I will debug into it.

**Comment by drauch** [comment]
I found the time to debug the problem in depth, for anybody ever running into the same issue:

An AJAX call in my application used a wrong ASP.NET pages handler, i.e., it called "./MyPage?handler=NotExistingHandler". To my surprise this doesn't return 404 or something similar but instead simply calls the OnGet default handler (why? I don't know yet, haven't found anything in the docs about it yet). Executing the default OnGet handler of course returned the whole page including the site.js inclusion at the end of the HTML body.

Then two jQuery surprises for me: jQuery.load(), for reasons described here https://github.com/jquery/jquery/issues/3975, loads scripts within loaded HTML documents with synchronous AJAX requests (hence the warning). Then also, jQuery document.ready handlers are executed when registered if the document.ready event has already been fired - which is the case, which is why I ran into handlers being registered multiple times and my hub connection being established a second time.

Sorry again for any inconveniences, it was a red herring that SignalR has anything to do with it.
